src/app/core/services/rundown-timing.service.ts

```diff
@@ -0,0 +1,125 @@
+import { BackwardRundownTiming, RundownTiming } from '../models/rundown-timing'
+import { RundownTimingType } from '../enums/rundown-timing-type'
+import { Rundown } from '../models/rundown'
+import { Segment } from '../models/segment'
+import { PartEntityService } from './models/part-entity.service'
+import { Part } from '../models/part'
+
+export class RundownTimingService {
+  private readonly partEntityService: PartEntityService = new PartEntityService()
+
+  public getExpectedStartEpochTime(rundownTiming: RundownTiming): number | undefined {
+    switch (rundownTiming.type) {
+      case RundownTimingType.FORWARD:
+        return rundownTiming.expectedStartEpochTime
+      case RundownTimingType.BACKWARD:
+        return this.getExpectedStartFromBackwardRundownTiming(rundownTiming)
+      default:
+        return undefined
+    }
+  }
+
+  private getExpectedStartFromBackwardRundownTiming(backwardRundownTiming: BackwardRundownTiming): number | undefined {
+    if (backwardRundownTiming.expectedStartEpochTime) {
+      return backwardRundownTiming.expectedStartEpochTime
+    }
+
+    if (backwardRundownTiming.expectedDurationInMs) {
+      return backwardRundownTiming.expectedEndEpochTime - backwardRundownTiming.expectedDurationInMs
+    }
+
+    return undefined
+  }
+
+  public getExpectedDurationInMs(rundownTiming: RundownTiming): number | undefined {
+    return rundownTiming.expectedDurationInMs
+  }
+
+  public getExpectedEndEpochTime(rundownTiming: RundownTiming): number | undefined {
+    switch (rundownTiming.type) {
+      case RundownTimingType.FORWARD:
+      case RundownTimingType.BACKWARD:
+        return rundownTiming.expectedEndEpochTime
+      default:
+        return undefined
+    }
+  }
+
+  private getAggregatedEndEpochTime(rundown: Rundown): number {
+    switch (rundown.timing.type) {
+      case RundownTimingType.FORWARD:
+        return rundown.timing.expectedEndEpochTime ?? rundown.timing.expectedStartEpochTime + this.getAggregatedDurationInMs(rundown)
+      case RundownTimingType.BACKWARD:
+        return rundown.timing.expectedEndEpochTime
+      case RundownTimingType.UNSCHEDULED:
+        return Date.now() + this.getRemainingRundownDuration(rundown)
+    }
+  }
+
+  private getAggregatedDurationInMs(rundown: Rundown): number {
+    return rundown.timing.expectedDurationInMs ?? rundown.segments.reduce((accumulatedDurationInMs, segment) => accumulatedDurationInMs + this.getSegmentDurationInMs(segment), 0)
+  }
+
+  private getSegmentDurationInMs(segment: Segment): number {
+    return segment.budgetDuration !== undefined
+      ? segment.budgetDuration
+      : segment.parts.reduce((accumulatedPartDuration: number, part: Part) => accumulatedPartDuration + this.getPartDuration(part), 0)
```

Comment:
Would this give the same result?
```
return segment.budgetDuration ?? segment.parts.reduce((accumulatedPartDuration: number, part: Part) => accumulatedPartDuration + this.getPartDuration(part), 0)
```

Reply:
This is changed in SOF-1680, but it might even be removed.